File: PredicttheWinner7.py
class Solution:
    def PredictTheWinner(self, nums: list) -> bool:
        def getScore(start, end):
            if start > end:
                return 0
            if not dp[start][end]:
                a = nums[start] + min(getScore(start + 1, end - 1), getScore(start + 2, end))
                b = nums[end] + min(getScore(start + 1, end - 1), getScore(start, end - 2))
                dp[start][end] = max(a, b)
            return dp[start][end]

        n = len(nums)
        dp = [[None] * n for _ in range(n)]
        return getScore(0, n - 1) * 2 >= sum(nums)


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.PredictTheWinner(nums))

File: PredicttheWinner7.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
        vector<int> dp(nums.size());
        for(int i = nums.size() - 1; i >= 0; i--) {
        	for(int j = i; j < nums.size(); j++) {
        		if(i == j) dp[j] = nums[i];
        		else dp[j] = max(nums[i] - dp[j], nums[j] - dp[j - 1]);
			}
		}
		
		return dp[nums.size() - 1] >= 0;
    }
    
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	Solution *solution = new Solution();
	solution->PredictTheWinner(nums) ? cout<<"true" : cout<<"false";
	
	return 0;
}
